code/modules/mob/living/say.dm

```diff
@@ -298,10 +298,81 @@ GLOBAL_LIST_INIT(department_radio_keys, list(
 	for(var/mob/M in listening)
 		if(M.client)
 			speech_bubble_recipients.Add(M.client)
+
 	var/image/I = image('icons/mob/talk.dmi', src, "[bubble_type][say_test(message)]", FLY_LAYER)
-	I.appearance_flags = APPEARANCE_UI_IGNORE_ALPHA
+
+	INVOKE_ASYNC(GLOBAL_PROC, /.proc/show_speech_text, message, message_language, src, speech_bubble_recipients, 40)
+
 	INVOKE_ASYNC(GLOBAL_PROC, /.proc/animate_speechbubble, I, speech_bubble_recipients, 30)
 
+/proc/show_speech_text(message, message_language, mob/living/L, var/list/show_to, duration)
+
+	if(!istype(L))
+		return
+
+	message = copytext(message, 1, 160) // no super long messages
+
+	var/datum/language/D = GLOB.language_datum_instances[message_language]
+	var/message_scrambled = D.scramble(message)
+
+
+	var/image/speech_text/S = new // create invisible object, bind it to speaking mob
+	S.loc = L
+	S.layer = FLY_LAYER
+
+	if(L.name != L.last_heard_name) // generate color based on name, skip if already generated
+		var/num = hex2num(copytext(md5(L.name), 1, 7))
+		L.last_used_color = hsv2rgb(num % 360, (num / 360) % 10 / 100 + 0.18, num / 360 / 10 % 15 / 100 + 0.85)

```

Comment:
Straight rip of https://github.com/goonstation/goonstation/blob/ac7bd846a4163a046b1ca4e0ae8f4636258825f5/code/mob/living.dm#L905-907

It's nice that you stole our exact values and hashing method, otherwise this might've not been as clear.